cmake/OpenCVDetectCUDA.cmake

```diff
@@ -355,6 +355,14 @@ if(CUDA_FOUND)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-Xptxas;-dlcm=ca")
   endif()
 
+  # Tell NVCC the maximum number of threads to be used to execute the compilation steps in parallel
+  # (option --threads was introduced in version 11.2)
+  if(NOT CUDA_VERSION VERSION_LESS "11.2")
+    if(NOT $ENV{CMAKE_BUILD_PARALLEL_LEVEL} STREQUAL "")
+      set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "--threads=$ENV{CMAKE_BUILD_PARALLEL_LEVEL}")
```

Comment:
The solution may break build with CMake before 3.12 as it does not define `CMAKE_BUILD_PARALLEL_LEVEL`.
Also if I use GNU Make, `cmake --build . -j4` will invoke 4 instances of nvcc during parallel build with make and each nvcc instance creates 4 jobs internally. I do not think that it's expected behaviour. The `--threads` makes more sense on Windows with msbuild. It runs nvcc sequentially and internal nvcc parallelism should reduce build time.

Reply:
I cannot see that checking the environment variable in an older version of CMake causes errors. I have tested CMake 3.1.3 and `$ENV{CMAKE_BUILD_PARALLEL_LEVEL}` evaluates to an empty string.

In our test environment the build time reduced significantly, but as you pointed out we were indeed using Windows. Do you have  any suggestions? Maybe add the option only if both the environment variable and `MSVC` are set?

Reply:
The first item may be ignored. I understood the code incorrectly.
MSVC check - yes, it's good idea.

Reply:
> > Also if I use GNU Make, `cmake --build . -j4` will invoke 4 instances of nvcc during parallel build with make and each nvcc instance creates 4 jobs internally. I do not think that it's expected behaviour. The `--threads` makes more sense on Windows with msbuild. It runs nvcc sequentially and internal nvcc parallelism should reduce build time.
> 
> Is this also true when using ninja on windows? If so should the check ensure that msbuild is being used not MSVC, i.e. something like `if (CMAKE_GENERATOR MATCHES "Visual Studio")`

It seems that it is not the case for Ninja on Windows. I've added the generator check.